packages/nuxt/src/core/templates.ts

```diff
@@ -121,14 +121,20 @@ export const schemaTemplate: NuxtTemplate<TemplateContext> = {
     })).filter(m => m.configKey && m.name && !adHocModules.includes(m.name))
 
     const relativeRoot = relative(resolve(nuxt.options.buildDir, 'types'), nuxt.options.rootDir)
+    const moduleMap: Record<string, any> = {}
 
     return [
       "import { NuxtModule } from '@nuxt/schema'",
       "declare module '@nuxt/schema' {",
       '  interface NuxtConfig {',
-      ...moduleInfo.filter(Boolean).map(meta =>
-        `    [${genString(meta.configKey)}]?: typeof ${genDynamicImport(meta.importName.startsWith('.') ? './' + join(relativeRoot, meta.importName) : meta.importName, { wrapper: false })}.default extends NuxtModule<infer O> ? Partial<O> : Record<string, any>`
-      ),
+      ...moduleInfo.filter(Boolean).map((meta) => {
+        const configKey = genString(meta.configKey)
+        const importName = meta.importName.startsWith('.') ? './' + join(relativeRoot, meta.importName) : meta.importName
+        moduleMap[importName] = configKey
+
+        return `    [${configKey}]?: typeof ${genDynamicImport(importName, { wrapper: false })}.default extends NuxtModule<infer O> ? Partial<O> : Record<string, any>`
+      }),
+      Object.keys(moduleMap).length > 0 ? `    modules: (NuxtModule | string | ${Object.keys(moduleMap).join(' | ')} | [NuxtModule | string, Record<string, any>] | ${Object.entries(moduleMap).map(([importName, configKey]) => `[${genString(importName)}, NuxtConfig[${genString(configKey)}]]`).join(' | ')})[]` : '',
```

Comment:
could we move the shared code (getting import name and config key) into the section above (L124)? that way we can consume it a bit more cleanly in both places below

Reply:
You will likely need to stringify the keys as well, something like:

```js
Object.keys(moduleMap).map(r => JSON.stringify(r)).join(' | ')
```

Reply:
I've made changes based on these suggestions.

> Actually, is there any reason we need these explicitly typed?
My answer to that would be 1) DX and 2) ensuring configurations match the desired module inputs to avoid errors.
Based on how developers intend to configure the `modules` property, all that Nuxt would do is provide IntelliSense for it - but I'm mostly making use of `NuxtConfig` interface itself!

In my opinion, it won't make a difference if someone configures to use absolute paths, because 1) they can continue passing options as a property than an array element or 2) if they are to use an array, at least options would be typed and it won't make difference between these two:

```ts
export default defineNuxtConfig({
  modules: [
   '/node_modules/.pnpm/@nuxtjs+color-mode@3.2.0/node_modules/@nuxtjs/color-mode/dist/module.mjs',
   // ['/node_modules/.pnpm/@nuxtjs+color-mode@3.2.0/node_modules/@nuxtjs/color-mode/dist/module.mjs', {}]
  ],
  colorMode: {},
})
```

The resulting schema would still include this as:

```ts
import { NuxtModule } from '@nuxt/schema'
declare module '@nuxt/schema' {
  interface NuxtConfig {
    ["colorMode"]?: typeof import("/node_modules/.pnpm/@nuxtjs+color-mode@3.2.0/node_modules/@nuxtjs/color-mode/dist/module.mjs").default extends NuxtModule<infer O> ? Partial<O> : Record<string, any>,
    // modules: (NuxtModule | string | "/node_modules/.pnpm/@nuxtjs+color-mode@3.2.0/node_modules/@nuxtjs/color-mode/dist/module.mjs" | [NuxtModule | string, Record<string, any>] | ["/node_modules/.pnpm/@nuxtjs+color-mode@3.2.0/node_modules/@nuxtjs/color-mode/dist/module.mjs", NuxtConfig['colorMode']])[]
  }
}
```

Reply:
Because `string` is a valid type of `modules` I don't think we'll get any benefit here by using a literal type, added to which you have to have added the string first, then saved the file, before you'll get auto-completion for the string you've just typed.

Reply:
That is a very fair point, and there is the point to realise that the schema validation occurs after the initial load.
I went for literal types because configurations would have to correspond to the modules - like we can't have the tailwind module options for color-mode.

Reply:
ah, but if we look at the below type:

```ts
modules: (
  NuxtModule |
  string |
  // here we do not need the literal type (and in fact it has no effect on type hinting because we also
  // allow string as a valid type)
  "@nuxtjs/color-mode" |
  [NuxtModule | string, Record<string, any>] |
  // here we _do_ need the literal type
  ["@nuxtjs/color-mode", NuxtConfig['colorMode']]
)[]
```

Reply:
Yes, I agree with you - that one is redundant. 🙂